pkg/controller/baremetalhost/baremetalhost_controller.go

```diff
@@ -110,6 +113,30 @@ type ReconcileBareMetalHost struct {
 	provisionerFactory provisioner.Factory
 }
 
+// Instead of passing a zillion arguments to the action of a phase,
+// hold them in a context
+type reconcileInfo struct {
+	log            logr.Logger
+	host           *metalkubev1alpha1.BareMetalHost
+	provisioner    provisioner.Provisioner
+	request        reconcile.Request
+	publisher      provisioner.EventPublisher
+	bmcCredsSecret *corev1.Secret
+}
+
+// Action for one step of reconciliation.
+//
+// - Return a result if the host should be saved and requeued without error.
+// - Return error if there was an error.
+// - Return double nil if nothing was done and processing should continue.
+type reconcileAction func(info reconcileInfo) (*reconcile.Result, error)
+
+// One step of reconciliation
+type reconcilePhase struct {
+	name   string
+	action reconcileAction
```

Comment:
FWIW I probably would have in-lined the type here, rather than define `reconcileAction`. It doesn't seem like you need it as a separate type, so you could condense it into what's below. But there's not any harm in breaking it out like you did, so feel free to go either way.

```
type reconcilePhase struct {
	name   string

        // - Return a result if the host should be saved and requeued without error.
        // - Return error if there was an error.
        // - Return double nil if nothing was done and processing should continue.
	action func(info reconcileInfo) (*reconcile.Result, error)
}
```